File: Algorithms/Backtracking/0099_recover_BST.py
#backtracking 

from typing import Optional 

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def recoverTree(self, root: Optional[TreeNode]) -> None:
        first = None
        second = None
        prev = None
        
        def inorder(node):
            nonlocal first, second, prev
            if not node:
                return

            inorder(node.left)
            
            if prev and prev.val > node.val:
                #first drop
                if not first:
                    first = prev
                    second = node 
                #second drop
                else:
                    second = node 
            
            #update prev
            prev = node
            
            inorder(node.right)
            
        inorder(root)
        
        #swap vals
        if first and second:
            first.val, second.val = second.val, first.val
        

File: Algorithms/Backtracking/0099_recover_BST.c
// Practice for: 0099_recover_BST.py
// Bridge logic from Python to C pointers here.

#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int val; 
    struct TreeNode* left; 
    struct TreeNode* right;
}; 

//need double pointer, pointer to a treeNode
void inorder(struct TreeNode* node, struct TreeNode** first, struct TreeNode** second, struct TreeNode** prev){
    if (node == NULL){
        return;
    }
    //use normal pointer to traverse 
    inorder(node->left, first, second, prev);
    //condition: prev val is bigger than the next (there are either 1 or 2 instants)
    if ((*prev) != NULL && (*prev)->val > node-> val){

        //first instance
        if((*first) == NULL){
            *first = *prev; 
            *second = node;
        } else {

            //second instance
            *second = node;
        }
    }
    *prev = node; 
    inorder(node->right, first, second, prev);
}

void recoverTree(struct TreeNode* root){
    struct TreeNode* first = NULL; 
    struct TreeNode* second = NULL; 
    struct TreeNode* prev = NULL;

    //swap vals here
    if (first != NULL && second != NULL){
        int temp = first->val; 
        first->val = second->val;
        second->val = temp;
    }
}